test/Dialect/FIRRTL/lower-domains.mlir

```diff
@@ -0,0 +1,86 @@
+// RUN: circt-opt --pass-pipeline='builtin.module(firrtl.circuit(firrtl-lower-domains))' --split-input-file %s | FileCheck --implicit-check-not firrtl.domain %s
+
+// Deeply check the lowering of domains and how associations are lowered.  Later
+// tests will ignore these deep checks.
+firrtl.circuit "Foo" {
+  // Two classes should be created:
+  //   1. A first that just represents the domain.  This has information that a
+  //      user needs to provide.
+  //   2. A second that returns this user-provided information and includes the
+  //      domain information.
+  // CHECK-LABEL: firrtl.class @ClockDomain
```

Comment:
I have questions about domains and classes and names -- you probably can guess, the usual kind.

Modulo making domains classlike, insofar as classes are nominally typed it may be important to give consideration here.  Probably the associations class is the main question (the other will continue as a domain (?) or at least can re-use the class name, assuming running with same namespace for them), while it's squarely compiler-generated it needs to be the same across SC units.

Something to think about.

Reply:
Yeah, good point. The re-use of the existing domain name for the "input" class works. The output class will need to have an ABI-determined name. While I think it's probably fine to keep this as `<name>_out` for now, this should likely be something better like `<name>$domainOut`. Or: something that uses `$` to prevent collision and is impossible to collide with. Same thing for the port name.

I won't treat this as blocking.